openedx/core/djangoapps/user_api/management/commands/bulk_user_org_email_optout.py

```diff
@@ -0,0 +1,122 @@
+"""
+One-off script to opt-out users for email from orgs.
+
+Input: A CSV file with a user_id,org pair per line. For example:
+
+1962921,FooX
+5506350,BarX
+5709986,FooX
+
+Opts-out every specified user/org combo rows from email by setting the 'email-optin' tag to 'False'.
+"""
+
+from __future__ import print_function
+import csv
+import logging
+import time
+
+from django.db import connections
+from django.core.management.base import BaseCommand, CommandError
+
+log = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """
+    Implementation of the bulk_user_org_email_optout command.
+    """
+    # Default number of user/org opt-outs to perform in each DB transaction.
+    DEFAULT_CHUNK_SIZE = 1000
+
+    # Default number of seconds to sleep between chunked user/org email opt-outs.
+    DEFAULT_SLEEP_BETWEEN_DELETES = 0.0
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--dry_run',
+            action='store_true',
+            help='Print proposed changes, but take no action.'
+        )
+        parser.add_argument(
+            '--chunk_size',
+            default=self.DEFAULT_CHUNK_SIZE,
+            type=int,
+            help='Maximum number of user/org opt-outs to perform in each DB transaction. Choose this value carefully to avoid DB outages!'
+        )
+        parser.add_argument(
+            '--sleep_between',
+            default=self.DEFAULT_SLEEP_BETWEEN_DELETES,
+            type=float,
+            help='Seconds to sleep between chunked opt-outs.'
+        )
+        parser.add_argument(
+            '--optout_csv_path',
+            help='Filepath to CSV file containing user/org email opt-outs.'
+        )
+
+
+    def handle(self, *args, **options):
+        """
+        Execute the command.
+        """
+        dry_run = options['dry_run']
+        chunk_size = options.get('chunk_size', self.DEFAULT_CHUNK_SIZE)
+        sleep_between = options.get('sleep_between', self.DEFAULT_SLEEP_BETWEEN_DELETES)
+        optout_path = options['optout_csv_path']
+
+        if chunk_size <= 0:
+            raise CommandError('Only positive chunk size is allowed ({}).'.format(chunk_size))
+        if sleep_between < 0:
+            raise CommandError('Only non-negative sleep between seconds is allowed ({}).'.format(sleep_between))
+
+        # Read the CSV file. Log the number of user/org rows read.
+        with open(optout_path, 'r') as csv_file:
+            optout_reader = csv.reader(csv_file)
+            optout_rows = list(optout_reader)
```

Comment:
I'm not sure what the size of this file is or the hardware it will be run on, but I assume fairly large? This ends up with two copies of it in memory for the whole run of the script. Might want to either read the file line by line or scope it so the csvreader doesn't stick around if you think it might be an issue.

Reply:
My recollection is something on the order hundreds of thousands of users, which means on the order of one million CSV rows, which means a file which is on the order of tens of megabytes.  That's pretty okay, definitely not ideal but this isn't going to stop anything.